python/python/bystro/supervised_ppca/tests/test_generative_pt.py

```diff
@@ -0,0 +1,80 @@
+import numpy as np
+import numpy.linalg as la
+import scipy.stats as st
+from ..gf_generative_pt import PPCApt, SPCApt, FactorAnalysispt
+
+def generate_data_ppca():
+    rng = np.random.default_rng(2021)
+    N = 10000
```

Comment:
I would be careful about the size of the test data here and throughout, especially when testing pytorch classes.  The most important cases to hit are 1 (to make sure we're not implicitly assuming we'll have more than one observation) and "a couple".  If the model has a different implementation for extremely large inputs (e.g. fancy matrix decomposition stuff) we'd want to check that too, but I don't think we need to worry about that here.

It's probably not as big a deal as `n_iterations`, which we definitely want to set explicitly in unit tests, but something to keep an eye on.

Reply:
If people try and fit a PCA model on a single observation they deserve it failing without explanation ;). The great thing about stochastic methods is that implementation is independent of sample size. And we would need at least 20k covariates for this to start being sad.

Reply:
1. Users never deserve anything failing without explanation.  Our job as experts in this particular domain is to build software that's usable by people who are not themselves experts, otherwise they'd just write this software themselves :)  If you feel differently about this we should probably discuss offline, because it's an important high-level design goal to sort out.
2. (Therefore) if our code makes assumptions about what inputs we'll get (which is fine), we need to check them upon receipt and fail informatively if those assumptions are violated.
3. There are many ways in a large, long-lived and evolving codebase that code could end up accepting inputs other than what the original author had in mind, ways that might not even be any one particular user or developer's fault.  The best defense against that inevitability is to make assumptions about input explicit.